src/test/java/htsjdk/samtools/util/IOUtilTest.java

```diff
@@ -520,4 +542,179 @@ public void testIsBlockCompressedOnJimfs(Path file, boolean checkExtension, bool
+    public void testCopyFileWriteException(final Path file) throws IOException {
+        final Path outFile = Files.createTempFile("tmp", ".tmp");
+        outFile.toFile().deleteOnExit();
+        outFile.toFile().setWritable(false);
+        IOUtil.copyFile(file.toFile(), outFile.toFile());
+    }
+
+    @DataProvider
+    public static Object[][] baseNameTests() {
+        return new Object[][]{
+                {TEST_DATA_DIR.resolve("ipsum.txt"), "ipsum"},
+                {TEST_DATA_DIR.resolve("ipsum.txt.bgz.wrongextension"), "ipsum.txt.bgz"},
+                {TEST_DATA_DIR.resolve("ipsum.txt.bgzipped_with_gzextension.gz"), "ipsum.txt.bgzipped_with_gzextension"},
+                {TEST_VARIANT_DIR.resolve("utils"), "utils"},
+                {TEST_VARIANT_DIR.resolve("not_real_file.txt"), "not_real_file"}
+        };
+    }
+
+    @Test(dataProvider = "baseNameTests")
+    public void testBasename(final Path file, final String expected) {
+        final String result = IOUtil.basename(file.toFile());
+        Assert.assertEquals(result, expected);
+    }
+
+    @DataProvider
+    public static Object[][] regExpTests() {
+        return new Object[][]{
+                {"\\w+\\.txt", new String[]{"5newline5.txt", "empty.txt", "ipsum.txt", "slurptest.txt"}},
+                {"^((?!txt).)*$", new String[]{"Level1.fofn", "Level2.fofn", "example.bam"}},
+                {"^\\d+.*", new String[]{"5newline5.txt"}}
+        };
+    }
+
+    @Test(dataProvider = "regExpTests")
+    public void testRegExp(final String regexp, final String[] expected) throws IOException {
+        final String[] allNames = {"5newline5.txt", "Level2.fofn", "example.bam", "ipsum.txt.bgz", "ipsum.txt.bgzipped_with_gzextension.gz", "slurptest.txt", "Level1.fofn", "empty.txt", "ipsum.txt", "ipsum.txt.bgz.wrongextension", "ipsum.txt.gz"};
+        final Path regExpDir = Files.createTempDirectory("regExpDir");
+        regExpDir.toFile().deleteOnExit();
+        final List<String> listExpected = Arrays.asList(expected);
+        final List<File> expectedFiles = new ArrayList<File>();
+        for (String name : allNames) {
+            final Path file = regExpDir.resolve(name);
+            file.toFile().deleteOnExit();
+            file.toFile().createNewFile();
+            if (listExpected.contains(name)) {
+                expectedFiles.add(file.toFile());
+            }
+        }
+        final File[] result = IOUtil.getFilesMatchingRegexp(regExpDir.toFile(), regexp);
+        Assert.assertEqualsNoOrder(result, expectedFiles.toArray());
+    }
+
+    @Test()
+    public void testReadLines() throws IOException {
+        final Path file = Files.createTempFile("tmp", ".txt");
+        file.toFile().deleteOnExit();
+        final int seed = 12394738;
+        final Random rand = new Random(seed);
+        final int nLines = 5;
+        final List<String> lines = new ArrayList<String>();
+        try (final BufferedWriter writer = Files.newBufferedWriter(file)) {

```

Comment:
Not sure it's nicer, but you can wrap the bufferedWriter with a PrintWriter and then call println() instead of write() + newLine().